File: android-native/app/src/main/java/in/videh/filedtracker/nativeapp/PunchInWindow.java
package in.videh.filedtracker.nativeapp;

import java.util.Calendar;
import java.util.Locale;
import java.util.TimeZone;

public final class PunchInWindow {
    private static final int START_MIN = 5 * 60;
    private static final int END_MIN = 13 * 60;

    private PunchInWindow() {}

    public static boolean isAllowed() {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("Asia/Kolkata"), Locale.US);
        int minutes = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE);
        return minutes >= START_MIN && minutes < END_MIN;
    }

    public static String blockedMessage() {
        return "Punch in is only allowed between 5:00 AM and 1:00 PM (IST).";
    }
}

File: ios-native/AAPAttendance/PunchInWindow.swift
import Foundation

/// Matches server `punchInWindow.ts`: punch-in from 7:00 AM IST through end of day
/// (including after 1:00 PM). Before 7:00 AM blocked unless unrestricted phone.
enum PunchInWindow {
    private static let startMinutes = 7 * 60
    private static let unrestrictedPhone = "9625692122"
    private static let ist = TimeZone(identifier: "Asia/Kolkata")!

    static func isUnrestrictedPhone(_ phone: String?) -> Bool {
        guard var digits = phone?.filter(\.isNumber), !digits.isEmpty else { return false }
        if digits.count > 10 { digits = String(digits.suffix(10)) }
        return digits == unrestrictedPhone
    }

    static func isAllowed(phone: String? = nil, now: Date = Date()) -> Bool {
        if isUnrestrictedPhone(phone ?? SessionStore.phone) { return true }
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = ist
        let hour = cal.component(.hour, from: now)
        let minute = cal.component(.minute, from: now)
        return hour * 60 + minute >= startMinutes
    }

    static var blockedMessage: String {
        "Punch in is allowed from 7:00 AM IST onward (including after 1:00 PM). Punch-in before 7:00 AM is not allowed."
    }
}
